client/src/game/reducers/cardDeductionReducer.ts

```diff
@@ -0,0 +1,131 @@
+  deck: readonly CardState[],
+  variant: Variant,
+  metadata: GameMetadata,
+) => {
+  const newDeck: CardState[] = deck.concat([]);
+  const cardCountMap = getCardCountMap(variant);
+
+  // if our variant shows played/discarded cards, then they're known by everyone
+  newDeck.filter((card) => !hands.some((hand) => hand.includes(card.order)))
+    .filter((card) => card.suitIndex !== null && card.rank !== null)
+    .forEach((card) => {
+      cardCountMap[card.suitIndex!][card.rank!] -= 1;
+    });
+
+  // We need to calculate our own hand first because those possibilities will be needed for
+  // pretending like we know what the other players see in our hand.
+  calculateHandPossibilities(hands[metadata.ourPlayerIndex], hands, newDeck, cardCountMap);
+  hands.filter((hand) => hand !== hands[metadata.ourPlayerIndex])
+    .forEach((hand) => { calculateHandPossibilities(hand, hands, newDeck, cardCountMap); });
+  return newDeck;
+};
+
+const calculateHandPossibilities = (
+  hand: readonly number[],
+  hands: ReadonlyArray<readonly number[]>,
+  deck: CardState[],
+  cardCountMap: readonly number[][],
+) => {
+  const remainingCardCountMap = Array.from(
+    cardCountMap,
+    (arr) => Array.from(arr),
+  );
+  const remainingCards: number[] = [];
+  const handCards: number[] = [];
+  hands.forEach((otherHand) => otherHand.forEach((o) => {
+    const deckCard = deck[o];
+    handCards.push(o);
+    // If card is not fully known then we need to run through the permutations.
+    // Otherwise, it's a card in another hand and we know we don't have that copy.
+    if (deckCard.suitIndex === null || deckCard.rank === null || otherHand === hand) {
+      remainingCards.push(o);
+    } else {
+      remainingCardCountMap[deckCard.suitIndex][deckCard.rank] -= 1;
+    }
+  }));
+  hand.forEach((o) => {
+    const card = deck[o];
+    const otherRemainingCards = remainingCards.filter((ro) => ro !== o).map((ro) => deck[ro]);
+    deck[o] = {
+      ...card,
+      possibleCardsFromDeduction: card.possibleCardsFromDeduction.filter(
+        (possibility) => possibilityValid(possibility, otherRemainingCards, remainingCardCountMap),
+      ),
+    };
+  });
+};
+
+const possibilityValid = (
+  [suit, rank]: readonly [number, number],
+  remainingCards: CardState[],
+  cardCountMap: readonly number[][],
+) => {
+  // Avoiding duplicating the map for performance, so trying to undo the mutation as we exit
+  cardCountMap[suit][rank] -= 1;
+  if (cardCountMap[suit][rank] >= 0 && remainingCardsValid(remainingCards, cardCountMap)) {
+    cardCountMap[suit][rank] += 1;
+    return true;
+  }
+  cardCountMap[suit][rank] += 1;
+  return false;
+};
+
+const remainingCardsValid = (
+  remainingCards: readonly CardState[],
+  cardCountMap: readonly number[][],
+) => remainingCards.length === 0 || remainingCards[0].possibleCardsFromDeduction.some(
+  (possibility) => possibilityValid(possibility, remainingCards.slice(1), cardCountMap),
```

Comment:
this is a pretty obtuse way to implement recursion, by creating 2 functions that call each other. I think a classic "base case/regular case" recursive function would be much cleaner.